pkg/virt-launcher/virtwrap/converter/numa_placement.go

```diff
@@ -0,0 +1,145 @@
+	numamap := map[uint32][]uint32{}
+	for _, tune := range cpuTune.VCPUPin {
+		cpu, err := strconv.ParseInt(tune.CPUSet, 10, 32)
+		if err != nil {
+			return nil, fmt.Errorf("expected only full cpu to be mapped, but got %v: %v", tune.CPUSet, err)
+		}
+		if _, exists := cpumap[uint32(cpu)]; !exists {
+			return nil, fmt.Errorf("vcpu %v is mapped to a not existing host cpu set %v", tune.VCPU, tune.CPUSet)
+		}
+		numamap[cpumap[uint32(cpu)].Id] = append(numamap[cpumap[uint32(cpu)].Id], tune.VCPU)
+	}
+	return numamap, nil
+}
+
+// numaMapping maps numa nodes based on already applied VCPU pinning. The sort result is stable compared to the order
+// of provided host numa nodes.
+func numaMapping(vmi *v1.VirtualMachineInstance, domain *api.DomainSpec, topology *cmdv1.Topology) error {
+	if topology == nil || len(topology.NumaCells) == 0 {
+		// If there is no numa topology reported, we don't do anything.
+		// this also means that emualted numa for e.g. memfd will keep intact
+		return nil
+	}
+	cpumap := cpuToCell(topology)
+	numamap, err := involvedCells(cpumap, domain.CPUTune)
+	if err != nil {
+		return fmt.Errorf("failed to generate numa pinning information: %v", err)
+	}
+
+	var involvedCellIDs []string
+	for _, cell := range topology.NumaCells {
+		if _, exists := numamap[cell.Id]; exists {
+			involvedCellIDs = append(involvedCellIDs, strconv.Itoa(int(cell.Id)))
+		}
+	}
+
+	domain.CPU.NUMA = &api.NUMA{}
+	domain.NUMATune = &api.NUMATune{
+		Memory: api.NumaTuneMemory{
+			Mode:    "strict",
+			NodeSet: strings.Join(involvedCellIDs, ","),
+		},
+	}
+
+	hugepagesSize, hugepagesUnit, hugepagesEnabled, err := hugePagesInfo(vmi, domain)
+	if err != nil {
+		return fmt.Errorf("failed to determine if hugepages are neabled: %v", err)
+	}
+
+	memory, err := QuantityToMebiByte(*getVirtualMemory(vmi))
+	if err != nil {
+		return fmt.Errorf("could not convert VMI memory to quantity: %v", err)
+	}
+	var mod uint64
+	if hugepagesEnabled {
+		cellCount := uint64(len(involvedCellIDs))
+		if memory < cellCount*hugepagesSize {
+			return fmt.Errorf("not enough memory requested to allocate at least one hugepage per numa node: %v < %v", memory, cellCount*(hugepagesSize*1024*1024))
+		} else if memory%hugepagesSize != 0 {
+			return fmt.Errorf("requested memory can't be divided through the numa page size: %v mod %v != 0", memory, hugepagesSize)
+		}
+		mod = (memory % (hugepagesSize * cellCount) / hugepagesSize)
+		if mod != 0 {
+			memory = memory - mod*hugepagesSize
+		}
+	}
+
+	virtualCellID := -1
+	for _, cell := range topology.NumaCells {
+		if vcpus, exists := numamap[cell.Id]; exists {
+			var cpus []string
+			for _, cpu := range vcpus {
+				cpus = append(cpus, strconv.Itoa(int(cpu)))
+			}
+			virtualCellID++
+
+			domain.CPU.NUMA.Cells = append(domain.CPU.NUMA.Cells, api.NUMACell{
+				ID:     strconv.Itoa(virtualCellID),
+				CPUs:   strings.Join(cpus, ","),
+				Memory: memory / uint64(len(numamap)),
```

Comment:
I'm not sure if we can do it this way.
There is no guarantee that there is enough memory in this numa node.
Until k8s 1.21 there is no mechanism afaik that checks numa memory allocation and even then it doesn't address our usecase.

I think we should decide on how much we can allocate per numa node based on what is available to us.
We can either parse [1] or query libvirt [2] ([3] for huge pages )

What do you think?

[1]  /sys/devices/system/node/[numa node X]/meminfo
[2] https://libvirt.org/html/libvirt-libvirt-host.html#virNodeGetFreeMemory
[3] https://libvirt.org/html/libvirt-libvirt-host.html#virNodeGetFreePages

Reply:
I was thinking about that (and I can do this to give better error messages), but in general my thinking was that qemu will in this case fail and virt-handler/virt-launcher will emit events and retry. With the check it may be handled more nicely, but at this stage I don't want to evict the VMI from the node forcefully, so I am not sure if I need a technically perfect detection of this scenario. I will add it if you think it is a minimal requirement.

Reply:
> I was thinking about that (and I can do this to give better error messages), but in general my thinking was that qemu will in this case fail and virt-handler/virt-launcher will emit events and retry. With the check it may be handled more nicely, but at this stage I don't want to evict the VMI from the node forcefully, so I am not sure if I need a technically perfect detection of this scenario. I will add it if you think it is a minimal requirement.

I don't think that qemu will fail, since we are not allocating all of this memory. I think it's the application in the guest will fail as we will present an incorrect memory distribution...

Reply:
> I don't think that qemu will fail, since we are not allocating all of this memory. I think it's the application in the guest will fail as we will present an incorrect memory distribution...

If that is the case, I am not sure if a memory check would help then. Since at startup time we can have enough memory, and later on not. Do you mostly thinking about making it less likely or do you see a way to fix the issue?

Reply:
Should we preallocate?

Reply:
Doing the pre-check and adding `<allocation mode="immediate"/>` (https://libvirt.org/formatdomain.html#memory-backing) may help.

Reply:
@vladikr as discussed, added the changes that the feature can only be used with hugepages and that the hugepages are immediately requested by qemu.